src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/InMemoryPartitionedAppTokenCacheAccessor.cs

```diff
@@ -0,0 +1,172 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Identity.Client.Cache;
+using Microsoft.Identity.Client.Cache.Items;
+using Microsoft.Identity.Client.Cache.Keys;
+using Microsoft.Identity.Client.Core;
+
+namespace Microsoft.Identity.Client.PlatformsCommon.Shared
+{
+    /// <summary>
+    /// Stores tokens for an application.
+    /// Partitions the access token collection by a key of client ID with tenant ID.
+    /// App metadata collection is not partitioned.
+    /// Refresh token, ID token, and account related methods are no-op.
+    /// </summary>
+    internal class InMemoryPartitionedAppTokenCacheAccessor : ITokenCacheAccessor
+    {
+        // perf: do not use ConcurrentDictionary.Values as it takes a lock
+        internal /* internal for test only */ readonly ConcurrentDictionary<string, ConcurrentDictionary<string, MsalAccessTokenCacheItem>> AccessTokenCacheDictionary =
+            new ConcurrentDictionary<string, ConcurrentDictionary<string, MsalAccessTokenCacheItem>>(1, 1);
+
+        internal /* internal for test only */  readonly ConcurrentDictionary<string, MsalAppMetadataCacheItem> AppMetadataDictionary =
+           new ConcurrentDictionary<string, MsalAppMetadataCacheItem>(1, 1);
+
+        protected readonly ICoreLogger _logger;
+
+        public InMemoryPartitionedAppTokenCacheAccessor(ICoreLogger logger)
+        {
+            _logger = logger ?? throw new System.ArgumentNullException(nameof(logger));
+        }
+
+        #region Add
+        public void SaveAccessToken(MsalAccessTokenCacheItem item)
+        {
+            var partitionKey = SuggestedWebCacheKeyFactory.GetClientCredentialKey(item.ClientId, item.TenantId);
+            string itemKey = item.GetKey().ToString();
+            // if a conflict occurs, pick the latest value
+            AccessTokenCacheDictionary
+                .GetOrAdd(partitionKey, new ConcurrentDictionary<string, MsalAccessTokenCacheItem>())[itemKey] = item;
+        }
+
+        public void SaveRefreshToken(MsalRefreshTokenCacheItem item)
+        {
```

Comment:
Should these throw?

Reply:
Smth like OperationNotSupportedException?

Reply:
GetAll methods are called during serialization. Other ones are not though.

Reply:
How about ?
- `OperationNotSupportedException` for GetItem / SaveItem 
- And OperationNotSupportedException if collection is not empty or Debug.Assert(collection is empty) for GetAllItems ?

Reply:
- Added comments
- Added `NotSupportedException` to Save, Get, Delete methods. 
  - N/A since we don't have collections for RT, IDT, Acc in app cache.